lib/internal/vm/Module.js

```diff
@@ -0,0 +1,150 @@
+  kErrored,
+} = internalBinding('module_wrap');
+
+const StatusMap = {
+  [kUninstantiated]: 'uninstantiated',
+  [kInstantiating]: 'instantiating',
+  [kInstantiated]: 'instantiated',
+  [kEvaluating]: 'evaluating',
+  [kEvaluated]: 'evaluated',
+  [kErrored]: 'errored',
+};
+
+let globalModuleId = 0;
+const kWrap = Symbol('wrap');
+const kContextModuleCount = Symbol('contextModuleCount');
+const kDependencyCache = Symbol('dependencyCache');
+
+class Module {
+  constructor(src, options = {}) {
+    let context;
+    if (options.context !== undefined && isContext(options.context.sandbox)) {
+      context = options.context.sandbox;
+      if (options.context[kContextModuleCount] === undefined)
+        throw new errors.TypeError('ERR_CONTEXT_NOT_MODULE_CONTEXT');
+    }
+
+    const url = options.url ||
+      `vm:module(${context !== undefined ?
+        options.context[kContextModuleCount]++ :
+        globalModuleId++})`;
+
+    const wrap = new ModuleWrap(src, url, {
+      context,
+      lineOffset: options.lineOffset,
+      columnOffset: options.columnOffset,
+    });
+
+    Object.defineProperties(this, {
+      [kWrap]: { value: wrap },
+      context: { value: options.context },
+    });
+  }
+
+  get status() {
+    return StatusMap[this[kWrap].getStatus()];
+  }
+
+  get url() {
+    return this[kWrap].url;
+  }
+
+  get namespace() {
+    if (this[kWrap].getStatus() !== kEvaluated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be evaluated');
+    return this[kWrap].namespace();
+  }
+
+  get dependencySpecifiers() {
+    if (this[kDependencyCache] !== undefined)
+      return this[kDependencyCache];
+
+    return this[kDependencyCache] = this[kWrap].getDependencySpecifiers();
+  }
+
+  async link(fn) {
+    const wrap = this[kWrap];
+    if (wrap.isLinked())
+      throw new errors.Error('ERR_VM_MODULE_ALREADY_LINKED');
+    if (wrap.getStatus() !== kUninstantiated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be uninstantiated');
+    const promises = [];
+    wrap.link(async (specifier) => {
+      const p = fn(specifier);
+      promises.push(p);
+      const m = await p;
+      if (!(m instanceof Module))
+        throw new errors.Error('ERR_VM_MODULE_NOT_MODULE');
+      return m[kWrap];
+    });
+    await Promise.all(promises);
+  }
+
+  instantiate() {
+    const wrap = this[kWrap];
+    if (wrap.getStatus() !== kUninstantiated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be uninstantiated');
+    if (!wrap.isLinked())
+      throw new errors.Error('ERR_VM_MODULE_UNLINKED');
+    this[kWrap].instantiate();
+  }
+
+  async evaluate(options) {
+    const status = this[kWrap].getStatus();
+    if (status !== kInstantiated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be instantiated');
+    if (status === kEvaluating ||
+      status === kEvaluated ||
+      status === kErrored)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'is already evaluated');
```

Comment:
There seems to be some issues with the logic.

- If it's `kInstantiated`, `evaluate()` should evaluate.
- If it's `kEvaluated`, reevaluating should do nothing.
- If it's `kErrored`, reevaluating should rethrow the error.
- If it's anything else, a `ERR_VM_MODULE_STATUS` should be thrown.

Tests to confirm these should be added.